### ISSUE TITLE
Poses Local Minima

### ISSUE THREAD

**Issue by xiaoyaolovlife** [WU-CVGL/BAD-Gaussians#26]
![image](https://github.com/user-attachments/assets/c895ec40-76b4-45e3-ac3d-a9d139fab7e9)
作者您好，非常优秀的工作，感谢你们做出的贡献！我有一个疑问：根据图片中的公式，我为了使L尽量小，有两个优化变量θ和T，在学习的过程中有没有可能陷入local minima，举个极端的例子，当所有virtual camera在优化过程中都聚集在一起了，这时候重建的结果就是一个模糊的场景，你们提出的无论是线性插值，cubic插值在某种程度上打破了这种情况的发生，但是如果我想用learning的思路生成virtual camera时就遇到了我所说的问题，请问你们有什么建议吗，感谢

"Dear authors, excellent work, and thank you for your contributions! I have a question: according to the formula in the image, in order to minimize L as much as possible, there are two optimization variables, θ and T. During the learning process, is it possible to get stuck in a local minimum? To give an extreme example, when all the virtual cameras converge together during optimization, the result of the reconstruction would be a blurry scene. The linear interpolation and cubic interpolation you proposed can, to some extent, prevent this situation from happening. However, if I want to use a learning-based approach to generate the virtual cameras, I encounter the problem I just described. Do you have any suggestions? Thank you."

**Comment by LingzheZhao** [comment]
您好,感谢您对我们工作的关注!

Local minima是有可能出现的：当相机的pose（轨迹）无法被准确优化出来的时候，3DGS场景为了去overfit各个多视角训练视图(在这里会有各种不同的模糊的pattern)，会在场景中产生很多高频的条纹状的3DGS，这时候我们会尝试去通过调低densification的阈值来降低场景中的3DGS数量，这时候会牺牲掉一些细节，但是3DGS就不容易往高频去overfit，迫使pose去往正确方向优化（这时候也可以一边观察`tensorboard`中pose的变化量，一边调整pose的学习率，让pose更快收敛）。

不过这些trick肯定是不优雅的，还有改进的空间。比如我们可能可以根据运动模糊的pattern（类似于2D的blur kernel），给camera trajectory提供一个初始估计；还有比如在连续拍摄的图片序列（视频）作为输入的时候，我们可以用一些运动学约束，比如速度、连续轨迹等，来更好地估计相机pose。

-----

Hello, thank you for your interest in our work!

Local minima is possible: when the camera pose (trajectory) cannot be accurately estimated, the 3DGS scene will generate many high-frequency striped 3DGS patterns in the scene in order to overfit the multi-view training views (here with various blurring patterns). And we will try to reduce the number of 3DGS in the scene by lowering the density threshold. Although some details are lost, but 3DGS will tend no to overfit to high frequencies, forcing the pose to optimize in the right direction. In the meantime, you can also watch the curve of delta-pose in `tensorboard` while adjusting the pose learning rate to make the pose converge faster.

However, these tricks are definitely not elegant and we can seek for some improvement. For example, we may be able to provide an initial estimate for the camera trajectory based on the motion blur pattern  (similar to the 2D blur kernel); and for example, when a sequence of continuously captured pictures (i.e. a video) is used as input, we can use some kinematic constraints, such as speed, continuous trajectory, etc., to better estimate the camera poses.

**Comment by xiaoyaolovlife** [comment]
感谢您的回复，“运动模糊的pattern（类似于2D的blur kernel），给camera trajectory提供一个初始估计”和“连续拍摄的图片序列（视频）作为输入的时候，我们可以用一些运动学约束”这两个思路我有想过，1. 但是无论从空域还是频域都无法从2D的模糊核中获取到z轴的运动信息吧，您有什么建议吗，2. 想在运动模糊视频中加入约束，比如速度和连续轨迹也存在少一个维度的问题吧

**Comment by xiaoyaolovlife** [comment]
另外我最开始想问的问题可能表述的有问题，您的回复我理解了，但是我遇到的情况并不是位姿优化的不够准确，产生了较多条纹状伪影（这个问题在用其他方法尝试的时候遇到过），而是生成virtual camera的过程中对于实际位姿的扰动矩阵波动的范围非常小，导致求得的所有的virtual camera都聚集在了一起，

第一张图是我自己的方法生成的所有virtual camera和对应的扰动矩阵
![image](https://github.com/user-attachments/assets/5f6ae044-170d-4dc4-af17-2bf1641510ce)
![image](https://github.com/user-attachments/assets/2a69dc4b-a316-42fd-9560-b51f783b714c)

----------------------------------------------------------------------------------
第二张图是用您的方法生成的所有virtual camera和对应的扰动矩阵
![image](https://github.com/user-attachments/assets/9691be67-4182-4a8c-bd7d-b0d500423986)
![image](https://github.com/user-attachments/assets/1ed41f4a-43fe-44f7-a2a6-9a3f7383f89b)

我看您代码关于 线性插值 位姿优化的思路是
1. 首先随机生成相对于实际相机坐标系下的两个控制点 cam_start和cam_end
2. 根据两个控制点在SE3中生成10个virtual camera
3. 根据实际采集的GT作为约束，反向优化每个virtual camera的位姿，但只保留cam_start和cam_end的优化结果，用作下一次插值使用

也就是说cam_start和cam_end是随着训练轮数的增加不断被优化的
有没有可能遇到一种情况，对于一张很模糊的图像（也就是说cam_start和cam_end理论应该在优化过程中离得非常远），cam_start和cam_end在优化的过程中却被优化到了离得非常近的区域，再进行插值，所有的virtual camera都聚集在一起，导致按这些视角渲染得到的图像非常模糊，影响重建的结果。如果出现了这样的原因，您觉得可以在参数上进行怎样的优化？

期待您的回复

**Comment by LingzheZhao** [comment]
不好意思回复晚了



> 感谢您的回复，“运动模糊的pattern（类似于2D的blur kernel），给camera trajectory提供一个初始估计”和“连续拍摄的图片序列（视频）作为输入的时候，我们可以用一些运动学约束”这两个思路我有想过，1. 但是无论从空域还是频域都无法从2D的模糊核中获取到z轴的运动信息吧，您有什么建议吗，2. 想在运动模糊视频中加入约束，比如速度和连续轨迹也存在少一个维度的问题吧

1. 嗯是的，这时候可能可以通过MVS/深度估计（两阶段），或3DGS训练过程（迭代优化）得到稠密的深度，把2d的估计lift到3d；
2. 对于轨迹约束，可以在3DGS迭代训练到过程中，将所有的相机位姿通过一条带连续性约束的轨迹去参数化，类似一些continuous-time SLAM或者trajectory estimation或者motion planning中的做法。

**Comment by LingzheZhao** [comment]
> 另外我最开始想问的问题可能表述的有问题，您的回复我理解了，但是我遇到的情况并不是位姿优化的不够准确，产生了较多条纹状伪影（这个问题在用其他方法尝试的时候遇到过），而是生成virtual camera的过程中对于实际位姿的扰动矩阵波动的范围非常小，导致求得的所有的virtual camera都聚集在了一起，
> 
> 第一张图是我自己的方法生成的所有virtual camera和对应的扰动矩阵 ![image](https://private-user-images.githubusercontent.com/170327167/386921512-5f6ae044-170d-4dc4-af17-2bf1641510ce.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE4Mjc0MDEsIm5iZiI6MTczMTgyNzEwMSwicGF0aCI6Ii8xNzAzMjcxNjcvMzg2OTIxNTEyLTVmNmFlMDQ0LTE3MGQtNGRjNC1hZjE3LTJiZjE2NDE1MTBjZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExN1QwNzA1MDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05OTAyYjAyYjI3ZTUzMTBmZGE1M2U1MDhkYjk4YzY4MDQwNmZhMzI4ZjVhZjc0OGM0ZjU1MTYyZTgzZDc2ODcwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bfxzw-Ly_s4JsMfrnK0yv2kXD-9OlVQjG1DDVFmAuag) ![image](https://private-user-images.githubusercontent.com/170327167/386921683-2a69dc4b-a316-42fd-9560-b51f783b714c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE4Mjc0MDEsIm5iZiI6MTczMTgyNzEwMSwicGF0aCI6Ii8xNzAzMjcxNjcvMzg2OTIxNjgzLTJhNjlkYzRiLWEzMTYtNDJmZC05NTYwLWI1MWY3ODNiNzE0Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExN1QwNzA1MDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wYTQwMTAzM2Y3ZjVlNjQzZWYwMjQ1ZjZmYzhjMWU3MDkzZDU4NDE3NTI0NzE2OTE0MTBjZjM2M2E0MzE3OGRhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.xvKMTzJxBmu_X9Hc40JNKMJVj4DoDZdxCiBrP6QhiiI)
> 
> 第二张图是用您的方法生成的所有virtual camera和对应的扰动矩阵 ![image](https://private-user-images.githubusercontent.com/170327167/386921541-9691be67-4182-4a8c-bd7d-b0d500423986.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE4Mjc0MDEsIm5iZiI6MTczMTgyNzEwMSwicGF0aCI6Ii8xNzAzMjcxNjcvMzg2OTIxNTQxLTk2OTFiZTY3LTQxODItNGE4Yy1iZDdkLWIwZDUwMDQyMzk4Ni5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExN1QwNzA1MDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yNTg5YWRlMGUzYWVlMGRiYWQ0MzBhNjA0NjA2Y2ZlZGM4ZmYzOTc3YTY2MzY0MjdjOThjYzU1MGYxNDM0Yjg1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.PEMw37EK1pdW7yeeCxxEceDxS5HpezOSmTjMXLiXAe4) ![image](https://private-user-images.githubusercontent.com/170327167/386921675-1ed41f4a-43fe-44f7-a2a6-9a3f7383f89b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE4Mjc0MDEsIm5iZiI6MTczMTgyNzEwMSwicGF0aCI6Ii8xNzAzMjcxNjcvMzg2OTIxNjc1LTFlZDQxZjRhLTQzZmUtNDRmNy1hMmE2LTlhM2Y3MzgzZjg5Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExN1QwNzA1MDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kYWVmMjhhODA5YTIzYzYwMjljMjdmNWZkZjUyODk5OGNlOWYyNWZmYzRjNDc3NWY3MmJlODM5Y2Q3NzliODNiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.CcLXC0Q-mBdlhdhk5yuOTQ2lHmJjFaHld-UexBuvCjQ)
> 
> 我看您代码关于 线性插值 位姿优化的思路是
> 
> 1. 首先随机生成相对于实际相机坐标系下的两个控制点 cam_start和cam_end
> 2. 根据两个控制点在SE3中生成10个virtual camera
> 3. 根据实际采集的GT作为约束，反向优化每个virtual camera的位姿，但只保留cam_start和cam_end的优化结果，用作下一次插值使用
> 
> 也就是说cam_start和cam_end是随着训练轮数的增加不断被优化的 有没有可能遇到一种情况，对于一张很模糊的图像（也就是说cam_start和cam_end理论应该在优化过程中离得非常远），cam_start和cam_end在优化的过程中却被优化到了离得非常近的区域，再进行插值，所有的virtual camera都聚集在一起，导致按这些视角渲染得到的图像非常模糊，影响重建的结果。如果出现了这样的原因，您觉得可以在参数上进行怎样的优化？
> 
> 期待您的回复

可以尝试提高pose学习率，或者降低图片分辨率/使用coarse-to-fine训练策略，后者可以降低pose优化的难度

**Comment by xiaoyaolovlife** [comment]
是这样的，我想用learning的思路想去学插值的位姿，简单地说，我在隐空间上生成的位姿扰动矩阵还是比较分散的，但想把扰动矩阵从隐空间映射到SE3空间下所有扰动矩阵就聚在一起了，根据我上面提到的理由，我觉得是因为位姿优化陷入了local minima，这种情况有什么建议吗

**Comment by LingzheZhao** [comment]
嗯 想问下这边“把扰动矩阵从隐空间映射到SE3空间”中的映射是怎样做的呢？在我们这里用的是机器人领域通用的Exp-Log映射(也就是用李代数来表示位姿扰动)，也有一些实现他们探索了不同的优化方式，例如[gsplat中就有一个用6d向量来表示旋转扰动并优化的demo](https://github.com/nerfstudio-project/gsplat/blob/dd66cbd597f376f103e90de8b6265087b482eac1/examples/utils.py#L117)

**Comment by GopiRajuMatta** [comment]
Please discuss in English so that it will help others also to resolve the similar issues they face..
